### ISSUE TITLE
thisType of callback parameter isn't enforced if the callback has another parameter

### ISSUE THREAD

**Issue by Arnavion** [Microsoft/TypeScript#10850]
**TypeScript Version:**  2.1.0-dev.20160910

**Code**

```ts
declare function foo(cb: (this: number, arg: string) => void): void;

foo(function (): void { this.length; }); // (1)
foo(function (arg: string): void { this.length; }); // (2)
```

`tsc -t es5 ./foo.ts`

**Expected behavior:**

Both (1) and (2) should complain about `length` not existing on `number`

**Actual behavior:** 
(1) correctly complains: `foo.ts(3,30): error TS2339: Property 'length' does not exist on type 'number'.` But there is no error for (2).

Adding `--noImplicitAny` doesn't change the result.

Adding `--noImplicitThis` gives `foo.ts(4,36): error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.` which confirms that `this` in (2) is getting inferred as `any`

**Comment by RyanCavanaugh** [comment]
Same root cause as #4241 - due to the `arg: string` type annotation, the entire function is not contextually typed. #4241 removes the restriction that type annotations in functions "turn off" contextual typing. Should be an easy change if you want to take a stab at it.

**Comment by Arnavion** [comment]
Ah, you're right. The `: string` annotation is unnecessary and removing it allows `this` to be inferred correctly. I didn't think to try that.
